// mup/lib/tasks/docker_service.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createService = createService;
exports.updateService = updateService;

var _utils = require('../utils');

function createService(taskList, _ref) {
  var image = _ref.image,
      name = _ref.name,
      publishedPort = _ref.publishedPort,
      targetPort = _ref.targetPort,
      replicas = _ref.replicas,
      envFile = _ref.envFile,
      hostname = _ref.hostname,
      bind = _ref.bind,
      log = _ref.log,
      volumes = _ref.volumes,
      docker = _ref.docker,
      networks = _ref.networks,
      hostLabels = _ref.hostLabels;

  taskList.executeScript('Start ' + name, {
    script: (0, _utils.resolvePath)(__dirname, 'assets/create-service.sh'),
    vars: {
      name: name,
      publishedPort: publishedPort,
      targetPort: targetPort,
      envFile: envFile,
      image: image,
      replicas: replicas,
      hostname: hostname
    }
  });
} /* eslint-disable */
function updateService() {}
